### PR TITLE
Add features to clean up unneeded packages

### PR DESCRIPTION
**Pull request by chanchiwai-ray**
- Add check-packages action to check if there are unneeded packages [0] can be removed and report how many disk spaces can be freed up.
- Add clean-up-packages action to remove unneeded packages [0] reported in the check-packages action.
- Enhance user experience for synchronize action
  - Remove indices (`mirror/**/dists/**/Packages`) before pulling from the upstream
  - Automatically clean up unneeded packages [0] during the action.

[0] unneeded packages means those packages that exist in the `mirror/**/pool/**/*.deb`, but have no reference in the packages indices `snapshot-*/**/dists/**/Packages*` and `mirror/**/dists/**/Packages`

### CODE REVIEW

**Review by jneo8 — APPROVED** [review]
The commits should be squash merge.
Other is good to me.

**Comment by chanchiwai-ray** [comment]
`PosixPath.readlink` only exists in `python >= 3.9`, https://docs.python.org/3/library/pathlib.html#pathlib.Path.readlink. This cause tests failed on focal

**Comment by esunar** [comment]
recheck

**Comment by chanchiwai-ray** [comment]
recheck

**Review by chanchiwai-ray — COMMENTED** [review]
recheck

File: actions.yaml [hunk 1]
```diff
@@ -22,3 +24,17 @@ publish-snapshot:
...
+    snapshots and the mirror path.
+clean-up-packages:
+  description: |
+    Remove the packages reported in the check-packages actions. You must
+    provide a parameter 'confirm=true' to actually perform the action.
+  params:
+    confirm:
+      description: "Confirm that you really want to perform this action."
+      type: boolean
+  required:
```
Comment (Pjack) [inline actions.yaml:43]: I do not read the code yet. So I have a question when I read this description
1. What happend when confirm = false?  
2. What can I do if I don't have enough confident to set confirm = true? 
We maybe should provide some suggestion to indicate how to perform the read-only mode ( is that check-packages ? )

File: src/charm.py [hunk 2]
```diff
@@ -117,14 +127,87 @@ def _on_config_changed(self, _):
...
+            event.set_results(
+                {"message": "Aborted! Please confirm your action with 'confirm=true'."}
+            )
+            return
+
+        start = time.time()
+        logger.info("Cleaning up unreferenced packages")
+        packages_to_be_removed, freed_up_space = self._check_packages()
+        for package in packages_to_be_removed:
+            package.unlink()
```
Comment (Pjack) [inline src/charm.py]: What if any exception happen here? I prefer to continue the job and clean up as much as file we can. And I prefer to add log here to record how which file is deleted.

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -453,15 +459,14 @@ def test_create_snapshot_action_strip_path(
...
         rand_n = random.randint(10, 100)
-        os_walk.return_value = a2g([rand_n])
         action_event = Mock()
         self.harness.charm._get_snapshot_name = Mock()
+        snapshot = Mock()
+        snapshot.name = "snapshot-{}".format(rand_n)
+        self.harness.charm._list_snapshots = Mock()
+        self.harness.charm._list_snapshots.return_value = [snapshot]
         self.harness.charm._on_list_snapshots_action(action_event)
         self.assertEqual(
```
Comment (Pjack) [inline tests/unit/test_charm.py]: What's the different if we use this function `assert_called_once_with`?
https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once_with

Comment (Pjack) [reply tests/unit/test_charm.py]: Do we need to test when there is no snapshot?  (return value = None or [] ? )

File: tests/unit/test_charm.py [hunk 4]
```diff
@@ -533,3 +538,81 @@ def test_publish_snapshot_action_fail(
...
+                    }
+                )
+            expected_snapshots = [
+                base_path / "snapshot-1970010{}".format(i) for i in range(3)
+            ]
+            for snapshot in expected_snapshots:
+                snapshot.mkdir(parents=True)
+            returned_snapshots = self.harness.charm._list_snapshots()
+            self.assertEqual(sorted(expected_snapshots), sorted(returned_snapshots))
+
```
Comment (Pjack) [inline tests/unit/test_charm.py:560]: Please add a test case when there is no snapshot.

File: tests/unit/test_charm.py [hunk 5]
```diff
@@ -533,3 +538,81 @@ def test_publish_snapshot_action_fail(
...
+            packages_to_be_removed,
+            freed_up_space,
+        ]
+        action_event = Mock()
+        self.harness.charm._on_check_packages_action(action_event)
+        self.assertEqual(
+            action_event.set_results.call_args,
+            call(
+                {
+                    "message": (
```
Comment (Pjack) [inline tests/unit/test_charm.py]: It doesn't make sense to limit the message in unit test to me.
Unit test is used to verify the specification of the function.
I would like to leave some flexibility for the implementation unless the format of message is part of specification and it will impact the caller.

assert_called_once may be good enough.

File: tests/unit/test_charm.py [hunk 6]
```diff
@@ -533,3 +538,81 @@ def test_publish_snapshot_action_fail(
...
+        packages_to_be_removed = [Mock() for i in range(3)]
+        self.harness.charm._check_packages = Mock()
+        self.harness.charm._check_packages.return_value = [
+            packages_to_be_removed,
+            freed_up_space,
+        ]
+
+        action_event = Mock(params={"confirm": False})
+        self.harness.charm._on_clean_up_packages_action(action_event)
+        self.assertEqual(
```
Comment (Pjack) [inline tests/unit/test_charm.py]: Ditto. 
What we need to verify is

1. parameter with "confirm"  => the result is good, unlink is called three times
2. parameter without "confirm" => unlink is not been called

Please try to think about 
1. how will you test the fucntion if you don't know the details of implementation.
2. If you want to refactor the code again, what is the pass criteria?

File: tests/unit/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+import utils
+
+
+class TestUtils(unittest.TestCase):
+    def test_find_packages_by_indices(self):
+        nums = 3
+        expected_outputs = set()
+        returned_output = set()
+        # Generate a test file structure for testing
+        with tempfile.TemporaryDirectory() as tmpdirname:
```
Comment (Pjack) [inline tests/unit/test_utils.py]: We can put the golden testing folder with the unit test together. 
Then we can
1. create many differet test environment easily
2. do not create the test environment by code which is complicated and may be buggy.

File: tests/unit/test_utils.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                    indices, base=archive_root
+                )
+
+        self.assertEqual(sorted(expected_outputs), sorted(returned_output))
+
+    def test_locate_package_indices(self):
+        nums = 3
+        expected_outputs = []
+        # Generate a test file structure for testing
+        with tempfile.TemporaryDirectory() as tmpdirname:
```
Comment (Pjack) [inline tests/unit/test_utils.py]: The code to generate temporary directory seems duplicated. We can try to put the testing envirnoment with the unit test together.

File: tests/unit/test_utils.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            returned_package_indices = sorted(returned_package_indices)
+            self.assertEqual(expected_outputs[i][0], returned_archive_roots)
+            self.assertEqual(expected_outputs[i][1], returned_package_indices)
+
+        non_existing_path = pathlib.Path(str(uuid4()))
+        self.assertRaises(
+            FileNotFoundError, utils.locate_package_indices, non_existing_path
+        )
+
+    def test_convert_bytes(self):
```
Comment (Pjack) [inline tests/unit/test_utils.py:42]: I just want to confirm if you know parametrize in pytest?
https://docs.pytest.org/en/6.2.x/parametrize.html

Your solution here is okay. But sometimes if the unit test cases are not complicated, we will put the static data in the code directly. It can avoid bug in the unit test itself.

File: tests/unit/test_utils.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        units = ["bytes", "KB", "MB", "GB", "TB"]
+        test_params = [
+            (1000 ** (i + 1), "{:.1f} {}".format(1000 ** (i + 1) / 1024 ** (i), unit))
+            for i, unit in zip(range(5), units)
+        ]
+        for num, expected in test_params:
+            with self.subTest():
+                self.assertEqual(utils.convert_bytes(num), expected)
+
+    def test__locate_packages_from_index(self):
```
Comment (Pjack) [inline tests/unit/test_utils.py:53]: two underline __  ?  I don't know how other people handle this. But I will use only 1 underline. Not a critical issue, just curious.

**Review by Pjack — CHANGES_REQUESTED** [review]

File: actions.yaml (hunk 1 above)
Comment (chanchiwai-ray) [reply actions.yaml:43]: 1. if confirm=false, the action will not be performed, and the action return a message saying that you should put confirm=true to actually perform the clean up
2. yes, it is expected that user should use check-packages action to view what packages (the path-to-packages will be return by the action) can be removed.

File: src/charm.py (hunk 2 above)
Comment (chanchiwai-ray) [reply src/charm.py]: I suspect that the action will fail at the point where error occurs. I think you are right, we should add some error handling and logs here .

File: tests/unit/test_charm.py (hunk 3 above)
Comment (chanchiwai-ray) [reply tests/unit/test_charm.py]: I did not change to `assert_called_once_with` since it was originally how is written, I think changing to `assert_called_once_with` is also okay. 

Agree, we should also test that because it is also not tested before.

File: tests/unit/test_charm.py (hunk 4 above)
Comment (chanchiwai-ray) [reply tests/unit/test_charm.py:560]: ACK!

File: tests/unit/test_charm.py (hunk 5 above)
Comment (chanchiwai-ray) [reply tests/unit/test_charm.py]: Thanks, that's indeed true. I will update that.

File: tests/unit/test_charm.py (hunk 6 above)
Comment (chanchiwai-ray) [reply tests/unit/test_charm.py]: Right, thanks, I should focus on the logic of the function rather than specific details in the implementation. I will update that.

File: tests/unit/test_utils.py (hunk 9 above)
Comment (chanchiwai-ray) [reply tests/unit/test_utils.py:42]: This is unittest way of doing parametrization, since `unittest.TestCase` does not work well with `pytest.mary.parametrize`. But I have not reason to stick with unittest if we choose pytest (I am just following the original author's choice).

File: tests/unit/test_utils.py (hunk 10 above)
Comment (chanchiwai-ray) [reply tests/unit/test_utils.py:53]: Right, I am not sure how is the convention here: normally we will name the test as `test_function_name` but this is a test of "private" function in `utils.py` which is called `_locate_packages_from_index()`, so I just keep the original name.

File: tests/unit/test_utils.py (hunk 7 above)
Comment (chanchiwai-ray) [reply tests/unit/test_utils.py]: I was also thinking about that, but since you requested it, let me just do it.

File: tests/unit/test_utils.py (hunk 8 above)
Comment (chanchiwai-ray) [reply tests/unit/test_utils.py]: same above

**Review by chanchiwai-ray — COMMENTED** [review]
recheck

File: src/charm.py [hunk 11]
```diff
@@ -117,14 +127,106 @@ def _on_config_changed(self, _):
...
+            try:
+                package.unlink()
+                logger.info("Removed {}".format(package))
+            except Exception as e:
+                logger.error(e)
+                success = False
+        elapsed = time.time() - start
+        logger.info("Clean up complete, took {}s".format(elapsed))
+
+        if not success:
```
Comment (esunar) [inline src/charm.py]: Can you revert if else

**Review by esunar — CHANGES_REQUESTED** [review]
Overall good, just a small change.

File: src/charm.py (hunk 11 above)
Comment (chanchiwai-ray) [reply src/charm.py]: yes

File: actions.yaml (hunk 1 above)
Comment (Pjack) [reply actions.yaml:43]: I will add the description in action.yaml

Comment (mkalcok) [reply actions.yaml:43]: I just want to add that this type of a "confirmation" parameter is common, for example a lot of openstack charms use `i-really-mean-it` that needs to be set to `true` for some of the potentially dangerous actions.

File: src/charm.py [hunk 12]
```diff
@@ -117,14 +127,101 @@ def _on_config_changed(self, _):
...
+            logger.info(
+                "clean up action not performed because the user did not confirm "
+                "the action."
+            )
+            event.set_results(
+                {"message": "Aborted! Please confirm your action with 'confirm=true'."}
+            )
+            return
+
+        start = time.time()
```
Comment (mkalcok) [inline src/charm.py:182]: Out of curiosity, is the timing of this action really important? All actions be default display `enqueued`, `started` and `completed` timestamps so user can get pretty god idea about how long the action execution took.

**Review by Pjack — APPROVED** [review]

File: src/utils.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+            if line.startswith("Filename: ")
+        }
+    return packages
+
+
+def _get_opener(path):
+    """Return appropriate opener to open an index file."""
+    extension = Path(path).suffix
+    if extension == ".gz":
+        from gzip import open as gopen
```
Comment (mkalcok) [inline src/utils.py:86]: Is there a chance that the system wont have all three of these packages (`bz2`/`gzip`/`lzma`)? If not I think they should be imported at the top of the module.

**Review by mkalcok — COMMENTED** [review]
Overall I think it's fine so I don't want to block this, but I left few inline comments.

**Review by esunar — APPROVED** [review]

File: src/utils.py (hunk 13 above)
Comment (chanchiwai-ray) [reply src/utils.py:86]: it is quite likely we don't have all three of them, this is mainly use for opening the "Package" file under `**/dists/**/Package*`. That file can be in any of these formats as mentioned in the debian package format documentation, and it up to the maintainer to choose which or all formats are in that directory.

File: src/charm.py (hunk 12 above)
Comment (chanchiwai-ray) [reply src/charm.py:182]: I also don't think it important. But I am saw that the original author did that in the synchronize action so I was thinking it might be useful  and just add that.

Comment (mkalcok) [reply src/charm.py:182]: +1 for consistency 😆

File: src/utils.py (hunk 13 above)
Comment (mkalcok) [reply src/utils.py:86]: but shouldn't the charm install all of them to be prepared to deal with whatever maintainer chooses? (both compression programs and their python bindings)

Comment (mkalcok) [reply src/utils.py:86]: Just to be more clear, my original question was more about the conditional imports. Why aren't we importing packages at top of the? Is there a chance that the system won't have some of these (python) packages available?

Bottom line is that I think that these import should be on top 😆

Comment (chanchiwai-ray) [reply src/utils.py:86]: Oh, I miss understood your question.. Yeah, I believe they exist in Python 3s' standard libaries. The only reason that I don't do that is that it's very likely that we don't use them all. But after a second thought, these modules is pretty small, it probably doesn't hurt to import them at the top.
